File: Problem-0001-Two-Sum/Problem-0001-Optimized/java/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

//=================================-Solution-=================================
public class Solution {
    //------------------------------Two-Sum-----------------------------------
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, List<Integer>> valueIndexMap = new HashMap<>();
        for (int index = 0; index < nums.length; index++) {
            if (!valueIndexMap.containsKey(nums[index])) {
                valueIndexMap.put(nums[index], new ArrayList<>());
            }
            valueIndexMap.get(nums[index]).add(index);
        }
        for (Map.Entry<Integer, List<Integer>> entry : valueIndexMap.entrySet()) {
            int primaryNum = entry.getKey();
            int remainderNum = target - primaryNum;
            if (valueIndexMap.containsKey(remainderNum)) {
                if (primaryNum == remainderNum) {
                    return new int[] {entry.getValue().get(0), entry.getValue().get(1)};
                } else {
                    return new int[] {entry.getValue().getFirst(), valueIndexMap.get(remainderNum).getFirst()};
                }
            }
        }
        return null;
    }
}

File: Problem-0001-Two-Sum/Problem-0001-Optimized/typescript/solution.ts
//=================================-Solution-=================================

//----------------------------------Two-Sum-----------------------------------
function twoSum(nums: number[], target: number): number[] {
    const map: Map<number, Array<number>> = new Map();
    for (let index: number = 0; index < nums.length; index++) {
        const value: number = nums[index];
        if (!map.has(value)) {
            map.set(value, new Array<number>());
        }
        const valueArray: Array<number> = map.get(value);
        valueArray.push(index)
        map.set(value, valueArray);
    }
    for (const [value, indexes] of map.entries()) {
        const remainderNum: number = target - value;
        if (map.has(remainderNum)) {
            switch (value) {
                case remainderNum:
                    if (indexes.length >= 2) {
                        return [indexes[0], indexes[1]];
                    }
                    break;
                default:
                    const secondaryIndexArray: number[] = map.get(remainderNum);
                    return [indexes[0], secondaryIndexArray[0]];
            }
        }
    }
    return null;
}